lib/sqlalchemy/orm/decl_api.py

```diff
@@ -1252,7 +1253,15 @@ def _resolve_type(
                 )
             else:
                 python_type_type = python_type_to_check.__origin__
-                search = ((python_type, python_type_type),)
+                if is_union(python_type_to_check):
```

Comment:
it's a little guessy here.   can we just run de-optoinalize on the types that we put into the type map as well?

Reply:
we're already trying to do this:

https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/lib/sqlalchemy/orm/decl_api.py#L1215-L1230

so the bug should be isolated to https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/lib/sqlalchemy/util/typing.py#L481-L500.    None should not be in the unions that are in the type map

Reply:
oh wow I find that decomposition of union args *extremely surprising*.

to use an example from the tests:

```python
type_annotation_map = {
   float | Decimal: Numeric(10, 2)
}
```

The design here seems to be that `Union` is used as a shortcut to say "If I use `Mapped[float]` or `Mapped[Decimal]` then please use `Numeric(10, 2)`".

The problem with that is the type doesn't make sense, `Numeric(asdecimal=True)` will *always* return `Decimal` and `Numeric(asdecimal=False)` will always return `float`.

In other words, a column with type `Numeric()` is never a union of `float | Decimal`. More formally, with this design there can be no union in the type annotation map which does not let you violate type safety. More concretely, `col: Mapped[float]` is "lying" to the type checker because you'd actually get a `Decimal` at runtime (or vice versa if the type annotation map had `float | Decimal: Numeric(10, 2, asdecimal=False)`).

Whereas a column of type `JSON()` actually is a union of `dict[str, Json] | list[Json] | str | int | float | bool | None`.

Using the `Json` union currently means the type annotation map ends up as this:

```python
type_annotation_map = {
    bool: JSON,
    float: JSON,
    int: JSON,
    str: JSON,
    dict[str, Json]: JSON,
    list[Json]: JSON,
    Json: JSON,
}
```

So the current design means an impossible union (the `Numeric` case) is possible to represent, but a possible one (any builtin or custom type whose `process_result_value` produces a union) is not possible to represent.

The SQLAlchemy 2.0 typing stuff is all fantastic, and I'm very thankful for all the work put into it but I have to say this union decomposition feels like it should be removed. One saving grace might be that union is not documented so far on https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#customizing-the-type-map

The type annotation map should mean that if I put `Mapped[str]`, I want a column type that will load a `str`, which by default is `String()`. And similarly if I put `Mapped[Json]` then I want a type that can load any of the values in that union.

I hope it's clear I am not trying to make a straw man argument out of the `float | Decimal` test case and that there's a real issue here. I'd be interested to know if there are any use cases I haven't thought of. I didn't find unions in the docs and the default `_type_map` does not use them.

I am willing to contribute to getting this fixed.

Reply:
> oh wow I find that decomposition of union args _extremely surprising_.
> 
> to use an example from the tests:
> 
> ```python
> type_annotation_map = {
>    float | Decimal: Numeric(10, 2)
> }
> ```
> 
> The design here seems to be that `Union` is used as a shortcut to say "If I use `Mapped[float]` or `Mapped[Decimal]` then please use `Numeric(10, 2)`".

no that's definitely not what it should do.  did you confirm it does that?    It only removes "None", it does not intend to decompose the whole Union

Reply:
> did you confirm it does that?

Yes. This is the script I used:

```python
from pprint import pprint
from typing import Union, Dict, List
from typing_extensions import TypeAlias

from sqlalchemy import JSON
from sqlalchemy.util import typing as compat_typing

_JsonPrimitive: TypeAlias = Union[str, int, float, bool, None]
_JsonObject: TypeAlias = Dict[str, "_Json"]
_JsonArray: TypeAlias = List["_Json"]
_Json: TypeAlias = Union[_JsonObject, _JsonArray, _JsonPrimitive]

type_annotation_map = {
    _Json: JSON,
}

# this dict comprehension matches
# https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/lib/sqlalchemy/orm/decl_api.py#L1223-L1229
pprint({
    sub_type: sqltype
    for typ, sqltype in type_annotation_map.items()
    for sub_type in compat_typing.expand_unions(
        typ, include_union=True, discard_none=True
    )
})
```

output:

```
{<class 'bool'>: <class 'sqlalchemy.sql.sqltypes.JSON'>,
 <class 'float'>: <class 'sqlalchemy.sql.sqltypes.JSON'>,
 <class 'int'>: <class 'sqlalchemy.sql.sqltypes.JSON'>,
 <class 'str'>: <class 'sqlalchemy.sql.sqltypes.JSON'>,
 typing.Dict[str, ForwardRef('_Json')]: <class 'sqlalchemy.sql.sqltypes.JSON'>,
 typing.List[ForwardRef('_Json')]: <class 'sqlalchemy.sql.sqltypes.JSON'>,
 typing.Union[typing.Dict[str, ForwardRef('_Json')], typing.List[ForwardRef('_Json')], str, int, float, bool, NoneType]: <class 'sqlalchemy.sql.sqltypes.JSON'>}
```

Reply:
see also these snippets from `test_typed_mapping.py::MappedColumnTest::test_unions`

https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/test/orm/declarative/test_typed_mapping.py#L1696

https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/test/orm/declarative/test_typed_mapping.py#L1722-L1723

https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/test/orm/declarative/test_typed_mapping.py#L1748-L1749

Reply:
that's just a component of the overall approach.  these types are definitely not decomposed in the final result, here's a script using the type that you claimed would not work, it is successfully differentiated against plain float and Decimal:

```py
from __future__ import annotations

from decimal import Decimal

from sqlalchemy import Numeric
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.sql.ddl import CreateTable


MyNumType = float | Decimal

class Base(DeclarativeBase):
    type_annotation_map = {
        MyNumType: Numeric(10, 2),

        Decimal: Numeric(15, 5),

    }

class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    d1: Mapped[MyNumType]
    d2: Mapped[float]
    d3: Mapped[Decimal]

print(CreateTable(A.__table__))
```

output:

```
CREATE TABLE a (
	id INTEGER NOT NULL, 
	d1 NUMERIC(10, 2) NOT NULL, 
	d2 FLOAT NOT NULL, 
	d3 NUMERIC(15, 5) NOT NULL, 
	PRIMARY KEY (id)
)

```

Reply:
then here's that same demo illustarting that `Optional` is successfully stripped out of the map

```py
from __future__ import annotations

from decimal import Decimal
from typing import Optional
from sqlalchemy import Numeric
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.sql.ddl import CreateTable


MyNumType = float | Decimal

class Base(DeclarativeBase):
    type_annotation_map = {
        MyNumType: Numeric(10, 2),

        Decimal: Numeric(15, 5),

    }

class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    d1: Mapped[MyNumType]
    d2: Mapped[MyNumType | None]
    d3: Mapped[Optional[MyNumType]]

print(CreateTable(A.__table__))
```

it's successfully detected in all cases

```py
CREATE TABLE a (
	id INTEGER NOT NULL, 
	d1 NUMERIC(10, 2) NOT NULL, 
	d2 NUMERIC(10, 2), 
	d3 NUMERIC(10, 2), 
	PRIMARY KEY (id)
)

```

Reply:
`is_union()` seems to treat `float | Decimal` and `Union[float, Decimal]` differently:

https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/lib/sqlalchemy/util/typing.py#L487

```python
from decimal import Decimal
from typing import Union

from sqlalchemy.util.typing import is_union

print(f"{is_union(Union[float, Decimal])=}")
print(f"{is_union(float | Decimal)=}")
```

```python
from __future__ import annotations

from decimal import Decimal
from typing import Union

from sqlalchemy import Numeric
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.sql.ddl import CreateTable


MyNumType = Union[float, Decimal]

class Base(DeclarativeBase):
    type_annotation_map = {
        MyNumType: Numeric(10, 2),

        Decimal: Numeric(15, 5),

    }

class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    d1: Mapped[MyNumType]
    d2: Mapped[float]
    d3: Mapped[Decimal]

print(CreateTable(A.__table__))
```

```
CREATE TABLE a (
	id INTEGER NOT NULL, 
	d1 NUMERIC(10, 2) NOT NULL, 
	d2 NUMERIC(10, 2) NOT NULL, 
	d3 NUMERIC(15, 5) NOT NULL, 
	PRIMARY KEY (id)
)
```

Reply:
it seems to do either of these things depending on if one sets `float | Decimal` vs. `Union[float, Decimal]`.

so we really can't just add a band aid here.   We need a full specification, lots more tests, and this likely has to be all 2.1 since we need to break backwards compat.  cc @CaselIT

Reply:
>  We need a full specification, lots more tests, and this likely has to be all 2.1 since we need to break backwards compat

I think we can merge this for 2.0 given that it was never specified what it should do.
Let's add unit tests to that part though

Reply:
btw the use of `expand_unions` there seems wrong.

Reply:
I really dont want to shove Optional in there at the last minute, I'd rather know why it's not working, noting that Optional is handled correctly for an example like this:

```py
from __future__ import annotations

from decimal import Decimal
from typing import Optional, Union
from sqlalchemy import Numeric
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.sql.ddl import CreateTable


# works
# MyNumType = float | Decimal

# also works
MyNumType = Union[float, Decimal]

class Base(DeclarativeBase):
    type_annotation_map = {
        MyNumType: Numeric(10, 2),

    }

class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    d1: Mapped[MyNumType]
    d2: Mapped[MyNumType | None]
    d3: Mapped[Optional[MyNumType]]

print(CreateTable(A.__table__))
```

None is stripped out.   adding it to the lookup should not have any effect

Reply:
These should have None stripped out:

```py
# does not work
#MyNumType = float | Decimal | None

# does not work
MyNumType = Union[float, Decimal, None]

```

I dont want to shove Optional in there at the last minute.   Need to know why existing routine is not working

Reply:
I agree that's the better approach than the one in this PR.

(And the reason it's not working is that expand_unions only removes None from the decomposed args, the first item of the tuple is still the original union).

---

It sounds like we're all on the same page about removing `expand_unions` from there, replacing it with `de_optionalize` for any unions.

Reply:
i mostly need to know if there was some original issue that had it working the way it does now for `Union` only and if there was thinking behind that

Reply:
I have thinkered a bit on this.

There is a bug in is_union.

Also we don't support at all type alias type (3.12+ `type X=int|str`).

I propose to fix the is_union to make this work. Then fix all the rest to make the 3.12 syntax also work in another change.

I can update the code of this change over the weekend